File: dynamic-programming/grid/minimum-falling-path-sum/ts/solution_dp.ts
function minFallingPathSum(matrix: number[][]): number {
    const m = matrix.length, n = matrix[0].length
    const dp: number[][] = Array.from({length: m}, (): number[] => Array(n + 2))
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n + 2; j++) {
            if (j === 0 || j === n + 1) {
                dp[i][j] = Number.MAX_SAFE_INTEGER
                continue
            }
            if (i === 0) {
                dp[0][j] = matrix[0][j - 1]
                continue
            }
            dp[i][j] = Math.min(dp[i - 1][j - 1], dp[i - 1][j], dp[i - 1][j + 1]) + matrix[i][j - 1]
        }
    }
    let minSum: number = Number.MAX_SAFE_INTEGER
    for (let i = 1; i < n + 1; i++) {
        minSum = Math.min(minSum, dp[m - 1][i])
    }
    return minSum
}

export { minFallingPathSum }

File: dynamic-programming/grid/minimum-falling-path-sum/java/Solution_Dp.java
class Solution_Dp {
    public int minFallingPathSum(int[][] matrix) {
        int m = matrix.length, n = matrix[0].length;
        int[][] dp = new int[m][n + 2];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n + 2; j++) {
                if (j == 0 || j == n + 1) {
                    dp[i][j] = Integer.MAX_VALUE;
                    continue;
                }
                if (i == 0) {
                    dp[0][j] = matrix[i][j - 1];
                    continue;
                }
                dp[i][j] = Math.min(Math.min(dp[i - 1][j - 1], dp[i - 1][j]), dp[i - 1][j + 1]) + matrix[i][j - 1];
            }
        }
        int minSum = Integer.MAX_VALUE;
        for (int i = 1; i < n + 1; i++) {
            minSum = Math.min(minSum, dp[m - 1][i]);
        }
        return minSum;
    }
}